### PR TITLE
Feature/password recovery

### LINKED ISSUE #118: Forgot Password
Users should be able to send an email to reset their password.

### PR DESCRIPTION
**Pull request by devnoll**
Added password recovery feature

### CODE REVIEW

File: app/controllers/users_controller.rb
Comment (ChristopherJMiller) [inline app/controllers/users_controller.rb:15]: Can you add tests for the SHOW and UPDATE routes to verify an unconfirmed account produces a status bad request?

**Review by ChristopherJMiller — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb
Comment (ConnorJC3) [reply app/controllers/users_controller.rb:15]: @ChristopherJMiller we have a test for that, this is just a fixup because the JSON was scuffed and that's harder to test. Will be tested in integration tests if we ever do that.

File: app/controllers/users_controller.rb
Comment (ConnorJC3) [inline app/controllers/users_controller.rb]: Hard to read, please convert to multiline if/else

File: app/models/user.rb
Comment (ConnorJC3) [inline app/models/user.rb:14]: Self note: CI needs fixing??

File: app/controllers/users_controller.rb
Comment (ConnorJC3) [inline app/controllers/users_controller.rb:62]: We either need a form that lets the user type in a new password, or we need to just generate a random one and give it to them (I'm in favor of the second here, but either works)

File: spec/factories/users.rb
Comment (ConnorJC3) [inline spec/factories/users.rb]: Dubious

**Review by ChristopherJMiller — DISMISSED** [review]
LGTM

File: app/controllers/users_controller.rb
Comment (ConnorJC3) [inline app/controllers/users_controller.rb]: Convert to `except`

**Review by ConnorJC3 — CHANGES_REQUESTED** [review]

**Comment by ConnorJC3** [comment]
Needs rebase

File: app/views/layouts/application.html.erb
Comment (ConnorJC3) [inline app/views/layouts/application.html.erb:8]: CD Note: Needs CSP fixup - consider self hosting?

File: app/views/users/password_reset.html.erb
Comment (ConnorJC3) [inline app/views/users/password_reset.html.erb]: Please convert all HTML files to 2 space tabs for consistency

File: spec/factories/users.rb
Comment (ConnorJC3) [inline spec/factories/users.rb]: Does this actually work :pepega: ???

File: app/controllers/users_controller.rb
Comment (ConnorJC3) [inline app/controllers/users_controller.rb:72]: `:password_reset` doesn't work here?

File: app/controllers/users_controller.rb
Comment (ConnorJC3) [inline app/controllers/users_controller.rb]: Unnecessary change?

File: app/views/users/password_reset.html.erb
Comment (ConnorJC3) [inline app/views/users/password_reset.html.erb]: Pre commit hooks skipped?
Self-note: needs major CI fixup

**Review by ConnorJC3 — CHANGES_REQUESTED** [review]
Prelim review, inserting forms has serious potential complexity and CSRF changes, needs thought/discussion before merging

File: app/controllers/users_controller.rb
Comment (devnoll) [reply app/controllers/users_controller.rb:62]: @ConnorJC3 decided that forms is a better approach.

File: app/models/user.rb
Comment (ConnorJC3) [reply app/models/user.rb:14]: Fixed in #169

File: app/views/users/password_reset.html.erb
Comment (ConnorJC3) [reply app/views/users/password_reset.html.erb]: Partially fixed in #169
Self-note: needs pre-commit fixup

**Comment by ConnorJC3** [comment]
@torresed needs rebase (ignore CI failure, will be fixed after rebase)

File: app/views/user_mailer/password_recovery.html.erb
Comment (ConnorJC3) [inline app/views/user_mailer/password_recovery.html.erb]: Minor, but I think I like "link below" better than "following link"

File: app/views/users/password_reset_confirmation.html.erb
Comment (ConnorJC3) [inline app/views/users/password_reset_confirmation.html.erb]: Can we change this to something like "Your password has been updated."? This version feels too informal.

File: app/views/users/password_reset_error.html.erb
Comment (ConnorJC3) [inline app/views/users/password_reset_error.html.erb]: Please use "email" not "e-mail".

File: config/routes.rb
Comment (ConnorJC3) [inline config/routes.rb]: Post request with token in the URL?

File: app/controllers/users_controller.rb
Comment (ConnorJC3) [inline app/controllers/users_controller.rb]: Can this tap be done via the update call instead

i.e. does `@user&.update(password_params, password_recovery_token: nil)` work?

File: app/controllers/application_controller.rb
Comment (ConnorJC3) [inline app/controllers/application_controller.rb:5]: Include this in the controller where it is used, because it probably won't be used elsewhere.

**Review by ConnorJC3 — CHANGES_REQUESTED** [review]
Minor changes, *do not merge* until CSP is fixed

File: app/controllers/application_controller.rb
Comment (devnoll) [reply app/controllers/application_controller.rb:5]: Must be included in ActionController.

**Comment by ConnorJC3** [comment]
CSP fixed: https://github.com/realliance/k8s-config/commit/cc6d97f6db6984d3e81498f3ec251d17c760679d

File: app/views/users/password_reset.html.erb
Comment (ConnorJC3) [inline app/views/users/password_reset.html.erb]: Minor nitpick but missing a space after `url:`

**Comment by ConnorJC3** [comment]
@torresed needs rebase, one minor nitpick, otherwise looks good :+1:

**Review by ConnorJC3 — APPROVED** [review]

**Pull request merged**
